### DISCUSSION TITLE
LottieView.animationDidFinish not called correctly when background behavior is set to continue playing

### DISCUSSION THREAD

**Discussion by arnopoulos** [airbnb/lottie-ios#2318] [General]
Hey everyone, I am currently experiencing an issue with the following code:
```swift
LottieView(animation: .named("test"))
    .playing()
    .backgroundBehavior(.continuePlaying)
    .animationDidFinish { completed in
        if completed {
            // Run code after completion
        }
    }
```

When backgrounding the app the animation continues to run however, the `animationDidFinish` handler is called as soon as the application is backgrounded rather than when the Lottie has completed running. In addition, when the Lottie has completed running it will not call the `animationDidFinish` handler, regardless of if the Lottie completes in the background/foreground.

I believe that this has something to do with how Core Animation treats `CAAnimationDelegate.animationDidStop(_:finished:)` and more specifically the problem seems to be in`Lottie.AnimationCompletionDelegate`.

## Which Version of Lottie are you using?

Lottie 4.4.0

## Expected Behavior 

When the background behavior is set to `LottieBackgroundBehavior.LottieBackgroundBehavior`, there are two things that should occur:
1. When backgrounding the app, the `animationDidFinish` handler should not be called
2. When the Lottie is finished regardless of application state the `animationDidFinish` handler should be called.

## Actual Behavior

When the background behavior is set to `LottieBackgroundBehavior.LottieBackgroundBehavior`, there are two things that occur:
1. When backgrounding the app, the `animationDidFinish` handler is called as if the Lottie was cancelled
2. When the Lottie is finished regardless of application state the `animationDidFinish` handler should is not called.

## Animation JSON

I don't have one that I can provide but I don't believe that matters here.

**Comment by calda** [comment]
Unfortunately I don't think it's possible for us to solve this for various reasons:
 1. If the animation completes while the app is in the background, we have no way of handling this. 
 2. When Core Animation calls `animationDidStop`, we don't know if this is because the animation was cancelled normally or if the app was backgrounded.

I'd suggest using `.backgroundBehavior(.pauseAndRestore)` if possible.

**Reply by arnopoulos** [reply]
Unfortunately not, that's not the behavior that we want. I'm not super familiar with the ins and outs of how this works would you be able to walk me through the architecture here? Because from a naive person like me I would think that we could have a some sort of timer set for when the background behavior is set to continue playing.
